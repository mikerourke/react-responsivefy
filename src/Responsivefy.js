import React from 'react';
import PropTypes from 'prop-types';

export default class Responsivefy extends React.Component {
  static propTypes = {
    height: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    margin: PropTypes.shape({
      top: PropTypes.number,
      bottom: PropTypes.number,
      left: PropTypes.number,
      right: PropTypes.number,
    }),
    children: PropTypes.node.isRequired,
  };

  static defaultProps = {
    margin: {
      top: 10,
      bottom: 10,
      left: 10,
      right: 10,
    },
  };

  constructor(props) {
    super(props);
    const { width, height } = this.props;
    this.aspect = +width / +height;
    this.state = {
      currentWidth: +width,
      currentHeight: +height,
    };
  }

  componentDidMount() {
    window.addEventListener('resize', this.resize);
    this.resize();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.resize);
  }

  resize = () => {
    const widthInPixels = window
      .getComputedStyle(this.ref.parentNode, null)
      .getPropertyValue('width');
    if (!widthInPixels) return;

    let updatedWidth = parseInt(widthInPixels, 10);
    let updatedHeight = Math.round(updatedWidth / this.aspect);

    const windowHeight = window.innerHeight;
    if (updatedHeight >= windowHeight) {
      updatedHeight = windowHeight;
      updatedWidth = Math.round(updatedHeight * this.aspect);
    }

    this.setState({ currentWidth: updatedWidth, currentHeight: updatedHeight });
  };

  handleRef = element => (this.ref = element);

  render() {
    const { height, width, margin } = this.props;
    const { left, right, top, bottom } = margin;
    const fullHeight = +height + +top + +bottom;
    const fullWidth = +width + +left + +right;

    return (
      <svg
        width={this.state.currentWidth}
        height={this.state.currentHeight}
        viewBox={`0 0 ${fullWidth} ${fullHeight}`}
        preserveAspectRatio="xMinYMid"
        ref={this.handleRef}
      >
        <g transform={`translate(${+left}, ${+top})`}>{this.props.children}</g>
      </svg>
    );
  }
}
