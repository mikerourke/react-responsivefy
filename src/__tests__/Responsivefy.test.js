import React from 'react';
import { render, mount } from 'enzyme';
import Responsivefy from '../Responsivefy';

const setup = (propOverrides, renderFn = render) => {
  const props = {
    height: 400,
    width: 600,
    margin: undefined,
    ...propOverrides,
  };

  const wrapper = renderFn(
    <Responsivefy {...props}>
      <rect width={100} height={100} y={10} x={12} />
    </Responsivefy>,
  );

  return { props, wrapper };
};

describe('Responsivefy Tests', () => {
  describe('when passed valid props', () => {
    test('it renders and matches snapshot when margins are specified', () => {
      const { wrapper } = setup({
        margin: { top: 20, bottom: 20, right: 20, left: 20 },
      });
      expect(wrapper).toMatchSnapshot();
    });

    test('it renders and matches snapshot when margins are not specified', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('when the componentDidMount event fires', () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    test('the resize EventListener is added', () => {
      const globalAddEventListener = global.addEventListener;
      const map = {};
      global.addEventListener = jest.fn().mockImplementation((event, cb) => {
        map[event] = cb;
      });

      setup({}, mount);

      global.addEventListener = globalAddEventListener;

      expect(typeof map.resize).toEqual('function');
    });

    const updateComputedStyleGetter = (value) => {
      const getPropertyValue = jest.fn().mockImplementation(name => value);
      global.getComputedStyle = jest.fn().mockImplementation((node, empty) => ({
        getPropertyValue,
      }));
    };

    test('the resize function fires', () => {
      updateComputedStyleGetter(600);
      const { wrapper } = setup({}, mount);

      const initialWidthFromState = wrapper.state('currentWidth');
      const newWidth = 900;
      global.window.innerWidth = newWidth;
      updateComputedStyleGetter(newWidth);

      global.window.dispatchEvent(new Event('resize'));

      const updatedWidthFromState = wrapper.state('currentWidth');

      expect(initialWidthFromState).not.toEqual(updatedWidthFromState);
      expect(updatedWidthFromState).toEqual(newWidth);
    });
  });

  describe('when the componentWillUnmount event fires', () => {
    afterAll(() => {
      jest.resetAllMocks();
    });

    test('the resize EventListener is removed', () => {
      const globalAddEventListener = global.addEventListener;
      const globalRemoveEventListener = global.removeEventListener;

      const map = {};
      global.addEventListener = jest.fn().mockImplementation((event, cb) => {
        map[event] = cb;
      });
      global.removeEventListener = jest.fn().mockImplementation((event, cb) => {
        map[event] = 'REMOVED';
      });

      const { wrapper } = setup({}, mount);
      wrapper.unmount();

      global.addEventListener = globalAddEventListener;
      global.removeEventListener = globalRemoveEventListener;

      expect(map.resize).toEqual('REMOVED');
    });
  });
});
