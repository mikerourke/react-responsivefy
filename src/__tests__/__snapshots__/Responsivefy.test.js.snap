// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Responsivefy Tests when passed valid props it renders and matches snapshot when margins are not specified 1`] = `
<svg
  height="400"
  preserveAspectRatio="xMinYMid"
  viewBox="0 0 620 420"
  width="600"
>
  <g
    transform="translate(10, 10)"
  >
    <rect
      height="100"
      width="100"
      x="12"
      y="10"
    />
  </g>
</svg>
`;

exports[`Responsivefy Tests when passed valid props it renders and matches snapshot when margins are specified 1`] = `
<svg
  height="400"
  preserveAspectRatio="xMinYMid"
  viewBox="0 0 640 440"
  width="600"
>
  <g
    transform="translate(20, 20)"
  >
    <rect
      height="100"
      width="100"
      x="12"
      y="10"
    />
  </g>
</svg>
`;
